File: MyLeetCode/python/Populating Next Right Pointers in Each Node.py
###
# Time Complexity: O(n)
# Space Complexity: O(n)
###
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if not root:
            return None
        q = collections.deque([root])
        while q:
            pre = None
            level_size = len(q)
            for i in range(level_size):
                cur = q.popleft()
                if pre:
                    pre.next = cur
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
                pre = cur
        return root

File: MyLeetCode/Leetcode/Populating Next Right Pointers in Each Node.cpp
#if 1 //recursive solution
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {

        if (!root)
            return;

        if (root->left && root->right)
        {
            root->left->next = root->right;

            if (root->next)
                root->right->next = root->next->left;
            else
                root->right->next = 0;

            connect(root->left);
            connect(root->right);
        }
    }
};
#elif 1 //iterative solution
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root) return;

        TreeLinkNode *next = 0;
        TreeLinkNode *pre = 0;
        TreeLinkNode *curr = root;

        while (curr)
        {
            if (!next)
                next = curr->left;
            if (pre)
                pre->next = curr->left;
            if (curr->left)
                curr->left->next = curr->right;

            pre = curr->right;
            curr = curr->next;
            if (!curr)
            {
                curr = next;
                next = pre = 0;
            }
        }
    }
};
#endif